Add plugin: Plugin REPL

talwrii opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

Link to my plugin: https://github.com/talwrii/plugin-repl

## Release Checklist
- [x] I have tested the plugin on
  - [ ]  Windows
  - [ ]  macOS
  - [x]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [ N/A] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x]  My project respects and is compatible with the original license of any code from other plugins that I'm using.
 - [x]     I have given proper attribution to these other projects in my `README.md`.

talwrii commented:
> #### Hello!
> 
> **I found the following issues in your plugin submission**
> 
> **Errors:**
> 
> ❌ Please don't include `Obsidian` in the plugin description
> 
> This check was done automatically. Do NOT open a new PR for re-validation. Instead, to trigger this check again, make a change to your PR and wait a few minutes, or close and re-open it.

This should be fixed

joethei commented:
To make the manual review process and future plugin maintenance more efficient, we recommend that you move all source code files into a `src` directory.

[this.app.workspace.on('active-leaf-change', () => { plugin.loadInit() })](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L90), 

[let vaultPath = this.app.vault.adapter.basePath](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L109)
You need to do a `instanceof FileSystemAdapter` and use the `getBasePath()` function instead of blindly adding the `@ts-ignore`

[id: 'repl-enter',](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L315), [id: 'repl-exec',](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L338), [id: 'repl-prompt-exec',](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L348)
No need to prefix this with `repl`, Obsidian already prefixes this with the plugin ID internally.
 
[app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L358), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L362), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L396), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L412), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L418), [editor: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L510), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L535), [app: any](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/main.ts#L536)
Don't use any here, the type is known.

[javascript](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/promptCommand.ts#L16), [Javascript](https://github.com/talwrii/plugin-repl/blob/7677b0b999aca0b785cdbb48fd5bb7d14fe7d2d2/promptCommand.ts#L21)
Please use the proper capitalization of this word, which is `JavaScript`

talwrii commented:
> To make the manual review process and future plugin maintenance more efficient, we recommend that you move all source code files into a src directory.
Done in:https://github.com/talwrii/plugin-repl/commit/2959a57bb474990d7dc657ad75727314845a48e1

> Events need to be registered
Done in: https://github.com/talwrii/plugin-repl/commit/7acf21f5446e60846c3ec7926b6097527d5fb410

> No need to prefix this with repl, Obsidian already prefixes this with the plugin ID internally.
Done in: https://github.com/talwrii/plugin-repl/commit/2da819581158f27b8c9410d83f03e63a4a8c334d

> Please use the proper capitalization of this word, which is JavaScript
Done in: https://github.com/talwrii/plugin-repl/commit/7c8b88b2e4841ae74b751a15fb6611f0947b00d8

> You need to do a instanceof FileSystemAdapter and use the getBasePath() function instead of blindly adding the @ts-ignore
> Don't use any here, the type is known.
Done in: https://github.com/talwrii/plugin-repl/commit/e1e1360f883ff35b32b2260dd7a4838ba34b5645
I've replaced most of these, sometimes with a change of the method I'm using.

Sometimes I'm using methods and properties (particularly related to plugins and commands) which aren't part of the documented API, which was part of the reason I was using any. I've implemented my own "PrivateApp" interface which adds these methods. 

Is there an analogue of `getBasePath` for [CapacitorApp](https://docs.obsidian.md/Reference/TypeScript+API/CapacitorAdapter). I'm mostly using this for my workaround `require` stuff.

talwrii commented:
I did a force push (after rebasing onto upstream master) because there were complaints here about a rebase merge being impossible.

talwrii commented:
I fixed the "as warnings". I'm still using "as casts" for some interfaces (specifically the wrappers I have for Templater and dataview) - the linting script Obsidian is using seems to allow those.

There does not seem to be a way to "test for" interfaces at runtime.

talwrii commented:
(did a force push because there was a warning about things no being rebaseable)

talwrii commented:
Huzzah. Thanks for all your help @joethei, I understand that you've been completely swamped with plugins :S .
